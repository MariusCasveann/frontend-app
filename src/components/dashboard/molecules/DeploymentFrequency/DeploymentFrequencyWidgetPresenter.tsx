import React from 'react';
import * as chartjs from 'chart.js';
import { Doughnut } from 'react-chartjs-2';
import { hasChartData, hasNoChartData } from '../../helpers/checkChartData';
import LoadingOrErrorPresenter from '../../../common/organisms/LoadingOrError/LoadingOrErrorPresenter';
import { CardLegend } from '../../../common/molecules/CardLegend/CardLegend';
import { colors } from '../../../../config/colors';
import './DeploymentFrequency.css';
import { MESSAGE_WHEN_NO_DATA } from '../../../../utils/constants';

interface DeploymentFrequencyProps {
    total: number;
    loading: boolean;
    formattedLabels: string[];
    error?: string;
    showTotalBuilds?: boolean;
    chartData?: chartjs.ChartData;
    messageWhenNoData?: string;
}

export default ({
    showTotalBuilds,
    chartData,
    messageWhenNoData,
    loading,
    error,
    formattedLabels,
    total
}: DeploymentFrequencyProps) => {
    if (loading || error) {
        return <LoadingOrErrorPresenter error={error} loading={loading} />;
    }

    if (hasChartData(chartData)) {
        return (
            <>
                {showTotalBuilds && (
                    <div className="builds-number">
                        <div className="total-builds">{total}</div>
                        <p>Total</p>
                    </div>
                )}
                {chartData && hasChartData(chartData) && <CardLegend labels={formattedLabels} colors={colors} />}
                {chartData && hasChartData(chartData) && (
                    <div className="canvas-container">
                        <Doughnut
                            data={chartData}
                            options={{
                                responsive: true,
                                legend: { display: false }
                            }}
                        />
                    </div>
                )}
            </>
        );
    }

    if (hasNoChartData(chartData)) {
        return <p>{messageWhenNoData || MESSAGE_WHEN_NO_DATA}</p>;
    }

    return null;
};
