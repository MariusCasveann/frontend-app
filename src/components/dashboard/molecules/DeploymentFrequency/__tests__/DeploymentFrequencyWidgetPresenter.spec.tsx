import * as React from 'react';
import renderer from 'react-test-renderer';
jest.mock('react-chartjs-2', () => ({ Doughnut: 'Doughnut' }));
jest.mock('../../../../common/organisms/LoadingOrError/LoadingOrErrorPresenter', () => 'LoadingOrErrorPresenter');
jest.mock('../../../../common/molecules/CardLegend/CardLegend', () => ({ CardLegend: 'CardLegend' }));
import DeploymentFrequencyWidgetPresenter from '../DeploymentFrequencyWidgetPresenter';

const mockData = {
    labels: ['Domain 1', 'Domain 2'],
    datasets: [
        {
            data: [1, 2],
            backgroundColor: ['#859ec4']
        }
    ]
};

describe('DeploymentFrequencyWidgetPresenter', () => {
    describe('when there are chart data', () => {
        it('renders Doughnut', () => {
            const wrapper = renderer
                .create(
                    <DeploymentFrequencyWidgetPresenter
                        showTotalBuilds={true}
                        total={3}
                        error=""
                        loading={false}
                        formattedLabels={mockData.labels}
                        chartData={mockData}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when chart data is undefined', () => {
        it('renders an error', () => {
            const wrapper = renderer
                .create(
                    <DeploymentFrequencyWidgetPresenter
                        showTotalBuilds={false}
                        error="Error"
                        total={3}
                        loading={false}
                        formattedLabels={mockData.labels}
                        chartData={undefined}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when chart data is undefined', () => {
        it('renders loading', () => {
            const wrapper = renderer
                .create(
                    <DeploymentFrequencyWidgetPresenter
                        showTotalBuilds={false}
                        error=""
                        total={3}
                        loading={true}
                        formattedLabels={mockData.labels}
                        chartData={undefined}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });
});
