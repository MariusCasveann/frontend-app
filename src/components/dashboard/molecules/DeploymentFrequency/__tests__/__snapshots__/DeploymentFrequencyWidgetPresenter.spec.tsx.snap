// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeploymentFrequencyWidgetPresenter when chart data is undefined renders an error 1`] = `
<LoadingOrErrorPresenter
  error="Error"
  loading={false}
/>
`;

exports[`DeploymentFrequencyWidgetPresenter when chart data is undefined renders loading 1`] = `
<LoadingOrErrorPresenter
  error=""
  loading={true}
/>
`;

exports[`DeploymentFrequencyWidgetPresenter when there are chart data renders Doughnut 1`] = `
Array [
  <div
    className="builds-number"
  >
    <div
      className="total-builds"
    >
      3
    </div>
    <p>
      Total
    </p>
  </div>,
  <CardLegend
    colors={
      Array [
        "#0c3d8a",
        "#859ec4",
        "#D8DADA",
        "#487ebd",
        "#007898",
        "#7fbbcb",
        "#58737a",
        "#92aab1",
        "#eee7a5",
        "#f1e364",
        "#f8b100",
        "#dcaeb9",
        "#ad405a",
        "#d9d4b9",
        "#a19351",
        "#886b22",
        "#7d7a58",
        "#564e02",
        "#c1dfc4",
        "#7dab6e",
        "#008d3a",
        "#36542c",
        "#263428",
      ]
    }
    labels={
      Array [
        "Domain 1",
        "Domain 2",
      ]
    }
  />,
  <div
    className="canvas-container"
  >
    <Doughnut
      data={
        Object {
          "datasets": Array [
            Object {
              "backgroundColor": Array [
                "#859ec4",
              ],
              "data": Array [
                1,
                2,
              ],
            },
          ],
          "labels": Array [
            "Domain 1",
            "Domain 2",
          ],
        }
      }
      options={
        Object {
          "legend": Object {
            "display": false,
          },
          "responsive": true,
        }
      }
    />
  </div>,
]
`;
